Zotero/Controllers/IdentifierLookupController.swift

```diff
@@ -0,0 +1,445 @@
+    private let disposeBag: DisposeBag
+    
+    private var lookupData: [LookupData] = []
+    private var lookupSavedCount = 0
+    private var lookupFailedCount = 0
+    private var lookupTotalCount: Int {
+        lookupData.count
+    }
+    private var lookupRemainingCount: Int {
+        lookupTotalCount - lookupSavedCount - lookupFailedCount
+    }
+    
+    internal weak var webViewProvider: IdentifierLookupWebViewProvider?
+    internal weak var presenter: IdentifierLookupPresenter? {
+        didSet {
+            guard presenter == nil, oldValue != nil else { return }
+            cleanupLookup(force: false, alreadyInQueue: false, update: Update(kind: .finishedAllLookups))
+        }
+    }
+    private var lookupWebViewHandlersByLookupSettings: [LookupWebViewHandler.LookupSettings: LookupWebViewHandler] = [:]
+    
+    // MARK: Object Lifecycle
+    init(
+        dbStorage: DbStorage,
+        fileStorage: FileStorage,
+        translatorsController: TranslatorsAndStylesController,
+        schemaController: SchemaController,
+        dateParser: DateParser,
+        remoteFileDownloader: RemoteAttachmentDownloader
+    ) {
+        self.fileStorage = fileStorage
+        self.dbStorage = dbStorage
+        self.translatorsController = translatorsController
+        self.schemaController = schemaController
+        self.dateParser = dateParser
+        self.remoteFileDownloader = remoteFileDownloader
+        
+        self.accessQueue = DispatchQueue(label: "org.zotero.IdentifierLookupController.accessQueue", qos: .userInteractive, attributes: .concurrent)
+        self.backgroundQueue = DispatchQueue(label: "org.zotero.IdentifierLookupController.backgroundProcessing", qos: .userInitiated)
+        self.observable = PublishSubject()
+        self.disposeBag = DisposeBag()
+        
+        setupObservers()
+    }
+    
+    // MARK: Actions
+    func initialize(libraryId: LibraryIdentifier, collectionKeys: Set<String>, completion: @escaping ([LookupData]?) -> Void) {
+        accessQueue.async(flags: .barrier) { [weak self] in
+            var lookupData: [LookupData]?
+            defer {
+                completion(lookupData)
+            }
+            guard let self = self else { return }
+            let lookupSettings = LookupWebViewHandler.LookupSettings(libraryIdentifier: libraryId, collectionKeys: collectionKeys)
+            if self.lookupWebViewHandlersByLookupSettings[lookupSettings] != nil {
+                lookupData = self.lookupData
+                return
+            }
+            inMainThread(sync: true) {
+                if let webView = self.webViewProvider?.addWebView() {
+                    let lookupWebViewHandler = LookupWebViewHandler(lookupSettings: lookupSettings, webView: webView, translatorsController: self.translatorsController)
+                    self.lookupWebViewHandlersByLookupSettings[lookupSettings] = lookupWebViewHandler
```

Comment:
`WKWebView` construction needs to happen in main thread, otherwise it crashes.

Reply:
Sure, but `lookupWebViewHandlersByLookupSettings` should be accessed/written from `accessQueue`. I guess in this case it doesn't matter since it's synchronous and inside `accessQueue` barrier block. I'd probably still keep web view initialization on main queue and assign it on `accessQueue`, but this shouldn't create any race conditions.

Reply:
Check https://github.com/zotero/zotero-ios/pull/707/commits/60b3c43d1a29b3e89e2967aad066e81f71b02bb0